akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -173,6 +174,217 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+   * responses in an order that doesn't directly match the consumed requests.
+   * For example, if two requests A and B enter the flow in that order the response for B might be produced before the
+   * response for A.
+   * In order to allow for easy response-to-request association the flow takes in a custom, opaque context
+   * object of type ``T`` from the application which is emitted together with the corresponding response.
+   */
+  def cachedHostConnectionPool[T](setup: HostConnectionPoolSetup)(implicit fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), HostConnectionPool] = {
+    val gateway = cachedGateway(setup)
+    clientFlow[T](setup.setup.settings)((_, _) ⇒ gateway).mapMaterialized(_ ⇒ HostConnectionPool(setup)(gateway))
+  }
+
+  /**
+   * Creates a new "super connection pool flow", which routes incoming requests to a (cached) host connection pool
+   * depending on their respective effective URI. Note that incoming requests must have either an absolute URI or
+   * a valid `Host` header.
+   *
+   * Since the underlying transport usually comprises more than a single connection the produced flow might generate
+   * responses in an order that doesn't directly match the consumed requests.
+   * For example, if two requests A and B enter the flow in that order the response for B might be produced before the
+   * response for A.
+   * In order to allow for easy response-to-request association the flow takes in a custom, opaque context
+   * object of type ``T`` from the application which is emitted together with the corresponding response.
+   */
+  def superPool[T](options: immutable.Traversable[Inet.SocketOption] = Nil,
+                   settings: HostConnectionPoolSettings = HostConnectionPoolSettings(system),
+                   defaultHeaders: immutable.Seq[HttpHeader] = Nil,
+                   log: LoggingAdapter = system.log)(implicit fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), Any] = {
+    val setup = ConnectionPoolSetup(encrypted = false, options, settings, defaultHeaders, log)
+    clientFlow[T](settings) { (request, userContext) ⇒
+      val effectiveRequest = request.withEffectiveUri(securedConnection = false)
+      val Uri.Authority(host, port, _) = effectiveRequest.uri.authority
+      cachedGateway(HostConnectionPoolSetup(host.toString(), port, setup))
+    }
+  }
+
+  /**
+   * Fires a single [[HttpRequest]] across the (potentially cached) host connection pool for the request's
+   * effective URI to produce a response future.
+   *
+   * Note that the request must have either an absolute URI or a valid `Host` header, otherwise
+   * the future will be completed with an error.
+   */
+  def singleRequest(request: HttpRequest,
+                    options: immutable.Traversable[Inet.SocketOption] = Nil,
+                    settings: HostConnectionPoolSettings = HostConnectionPoolSettings(system),
+                    defaultHeaders: immutable.Seq[HttpHeader] = Nil,
```

Comment:
Actually, looking at it a bit more closely, there is a reason for having this here.
Not such much to supply additional headers to the request (which the user could easily do on her side before the call), but rather because the `defaultHeaders` are part of the `ConnectionPoolSetup` and as such the cache key.
If we don't give the user the ability to supply default headers here they cannot target a host pool that was configured with `defaultHeaders` before.

Plus, the additional parameter doesn't really hurt anyone here, so I'd be inclined to keep it after all.